File: code/backend/java/core/src/main/java/games/paths/core/service/security/CsrfTokenService.java
package games.paths.core.service.security;

import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.security.MessageDigest;
import java.util.Base64;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * CsrfTokenService - Step 41: a stateless CSRF token bound to the access token it was issued with.
 * The token is HMAC-SHA256(secret, "csrf:" + accessToken), so nothing is stored and every backend agrees.
 */
public class CsrfTokenService {

    public static final String HEADER = "X-CSRF-TOKEN";
    private static final String ALGORITHM = "HmacSHA256";
    private static final String PREFIX = "csrf:";

    private final byte[] secret;
    private final boolean enforced;

    public CsrfTokenService(String secret, boolean enforced) {
        if (secret == null || secret.isBlank()) {
            throw new IllegalArgumentException("CSRF secret must not be blank");
        }
        this.secret = secret.getBytes(StandardCharsets.UTF_8);
        this.enforced = enforced;
    }

    public boolean isEnforced() {
        return enforced;
    }

    /** The token a client must echo back in {@value #HEADER} while it holds this access token. */
    public String tokenFor(String accessToken) {
        if (accessToken == null || accessToken.isBlank()) {
            return null;
        }
        try {
            Mac mac = Mac.getInstance(ALGORITHM);
            mac.init(new SecretKeySpec(secret, ALGORITHM));
            byte[] digest = mac.doFinal((PREFIX + accessToken.trim()).getBytes(StandardCharsets.UTF_8));
            return Base64.getUrlEncoder().withoutPadding().encodeToString(digest);
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("HMAC unavailable", e);
        }
    }

    /** Constant-time comparison of what the client sent against what this access token deserves. */
    public boolean matches(String accessToken, String presented) {
        String expected = tokenFor(accessToken);
        if (expected == null || presented == null || presented.isBlank()) {
            return false;
        }
        return MessageDigest.isEqual(expected.getBytes(StandardCharsets.UTF_8),
                presented.trim().getBytes(StandardCharsets.UTF_8));
    }
}

File: code/backend/python/app/core/services/security/csrf_token_service.py
"""Step 41 — a stateless CSRF token bound to the access token it was issued with (v0.37.7).

``token_for(access_token)`` is base64url(HMAC-SHA256(secret, "csrf:" + access_token)) with no
padding, exactly as the Java and AWS backends compute it, so nothing is stored anywhere.
"""
import base64
import hashlib
import hmac
from typing import Optional

HEADER = "X-CSRF-TOKEN"
_PREFIX = "csrf:"


class CsrfTokenService:
    def __init__(self, secret: str, enforced: bool = True):
        if not secret or not secret.strip():
            raise ValueError("CSRF secret must not be blank")
        self._secret = secret.encode("utf-8")
        self._enforced = bool(enforced)

    @property
    def enforced(self) -> bool:
        return self._enforced

    def token_for(self, access_token: Optional[str]) -> Optional[str]:
        """The token a client must echo back in X-CSRF-TOKEN while it holds this access token."""
        if not access_token or not access_token.strip():
            return None
        digest = hmac.new(self._secret, (_PREFIX + access_token.strip()).encode("utf-8"),
                          hashlib.sha256).digest()
        return base64.urlsafe_b64encode(digest).rstrip(b"=").decode("ascii")

    def matches(self, access_token: Optional[str], presented: Optional[str]) -> bool:
        """Constant-time comparison of what the client sent against what this token deserves."""
        expected = self.token_for(access_token)
        if expected is None or not presented or not presented.strip():
            return False
        return hmac.compare_digest(expected, presented.strip())
